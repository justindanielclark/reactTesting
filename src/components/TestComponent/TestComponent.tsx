import React, { useState } from "react";

type Props = {
  testing?: boolean;
  initialState?: number;
  heading?: string;
  listContent?: string[];
};

function TestComponent({ heading, listContent, initialState, testing }: Props) {
  const [state, setState] = useState(initialState ? initialState : 0);
  const handleClickIncrementButton = (e: React.MouseEvent): void => {
    setState(state + 1);
  };
  const handleClickDecrementButton = (e: React.MouseEvent): void => {
    setState(state - 1);
  };
  const renderList = (listContent: string[]): JSX.Element => {
    return (
      <ul>
        {listContent.map((item, idx) => {
          return <li key={idx}>{item}</li>;
        })}
      </ul>
    );
  };
  return (
    <div data-testid={testing ? "TestComponentRoot" : undefined}>
      <h1>{heading ? heading : "Default Heading"}</h1>
      {listContent ? renderList(listContent) : <p>No Data Provided</p>}
      <div>
        <p>State Values:</p>
        <p role={"log"}>{state}</p>
      </div>
      <button
        role="button"
        type="button"
        name="incrementBtn"
        onClick={handleClickIncrementButton}
        className="bg-green-900 p-1 rounded mx-2"
      >
        Increment
      </button>
      <button
        role="button"
        type="button"
        name="decrementBtn"
        onClick={handleClickDecrementButton}
        className="bg-red-900 p-1 rounded mx-2"
      >
        Decrement
      </button>
    </div>
  );
}

export default TestComponent;
