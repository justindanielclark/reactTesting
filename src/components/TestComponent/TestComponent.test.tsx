// import React from "react";
import TestComponent from "./TestComponent";
import { fireEvent, render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import userEvent from "@testing-library/user-event";

const testListItems = ["Justin", "Natasha", "Steve"];

function setup(jsx: JSX.Element) {
  return {
    user: userEvent.setup(),
    ...render(jsx),
  };
}

test("<TestComponent/> : Snapshot Comparison", () => {
  setup(<TestComponent testing />);
  const element = screen.getByTestId("TestComponentRoot");
  expect(element).toMatchSnapshot();
});

test("<TestComponent/> : Snapshot with Custom Header", async () => {
  const { baseElement } = setup(
    <TestComponent heading="Custom Header" testing />
  );
  expect(baseElement).toMatchSnapshot();
});

test("<TestComponent/> : Snapshot with ListItems", async () => {
  const { baseElement, getByRole } = setup(
    <TestComponent testing listContent={testListItems} />
  );
  expect(baseElement).toMatchSnapshot();
});

test("<TestComponent/> : Increments when Increment Button is Clicked", async () => {
  const { user, getByRole } = setup(<TestComponent />);
  const logElement = getByRole("log");
  const incrementBtn = getByRole("button", { name: "Increment" });
  await user.click(incrementBtn);
  await user.click(incrementBtn);
  expect(logElement.textContent).toBe("2");
});

test("<TestComponent/> : Decrements when Decrement Button is Clicked", async () => {
  const { user, getByRole } = setup(<TestComponent />);
  const logElement = getByRole("log");
  const decrementBtn = getByRole("button", { name: "Decrement" });
  await user.click(decrementBtn);
  await user.click(decrementBtn);
  expect(logElement.textContent).toBe("-2");
});
