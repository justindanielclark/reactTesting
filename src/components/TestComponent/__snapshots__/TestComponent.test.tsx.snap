// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TestComponent/> : Snapshot Comparison 1`] = `
<div
  data-testid="TestComponentRoot"
>
  <h1>
    Default Heading
  </h1>
  <p>
    No Data Provided
  </p>
  <div>
    <p>
      State Values:
    </p>
    <p
      role="log"
    >
      0
    </p>
  </div>
  <button
    class="bg-green-900 p-1 rounded mx-2"
    name="incrementBtn"
    role="button"
    type="button"
  >
    Increment
  </button>
  <button
    class="bg-red-900 p-1 rounded mx-2"
    name="decrementBtn"
    role="button"
    type="button"
  >
    Decrement
  </button>
</div>
`;

exports[`<TestComponent/> : Snapshot with Custom Header 1`] = `
<body>
  <div>
    <div
      data-testid="TestComponentRoot"
    >
      <h1>
        Custom Header
      </h1>
      <p>
        No Data Provided
      </p>
      <div>
        <p>
          State Values:
        </p>
        <p
          role="log"
        >
          0
        </p>
      </div>
      <button
        class="bg-green-900 p-1 rounded mx-2"
        name="incrementBtn"
        role="button"
        type="button"
      >
        Increment
      </button>
      <button
        class="bg-red-900 p-1 rounded mx-2"
        name="decrementBtn"
        role="button"
        type="button"
      >
        Decrement
      </button>
    </div>
  </div>
</body>
`;

exports[`<TestComponent/> : Snapshot with ListItems 1`] = `
<body>
  <div>
    <div
      data-testid="TestComponentRoot"
    >
      <h1>
        Default Heading
      </h1>
      <ul>
        <li>
          Justin
        </li>
        <li>
          Natasha
        </li>
        <li>
          Steve
        </li>
      </ul>
      <div>
        <p>
          State Values:
        </p>
        <p
          role="log"
        >
          0
        </p>
      </div>
      <button
        class="bg-green-900 p-1 rounded mx-2"
        name="incrementBtn"
        role="button"
        type="button"
      >
        Increment
      </button>
      <button
        class="bg-red-900 p-1 rounded mx-2"
        name="decrementBtn"
        role="button"
        type="button"
      >
        Decrement
      </button>
    </div>
  </div>
</body>
`;
